src/gbastd/types/rectangle.h

```diff
@@ -0,0 +1,34 @@
+#ifndef RECTANGLE_H
+#define RECTANGLE_H
+
+#include "u.h"
+#include "vector.h"
+
+typedef struct {
+    ivec2 lowerLeftPoint;
+    int height;
+    int width;
+} Rectangle;
+
+int checkCollision(Rectangle* rect1, Rectangle* rect2) {
+    if (rect1->lowerLeftPoint.x <= rect2->lowerLeftPoint.x + rect2->width &&
+    rect1->lowerLeftPoint.x + rect1->width >= rect2->lowerLeftPoint.x &&
+     rect1->lowerLeftPoint.y <= rect2->lowerLeftPoint.y + rect2->height &&
+      rect1->lowerLeftPoint.y + rect1->height >= rect2->lowerLeftPoint.y) {
+        return 1;
+      }
+
+    return 0;
+}
+
+int movingObjectsCollision(Rectangle rect1, Rectangle rect2, ivec2 rect1MovementDirection, ivec2 rect2MovementDirection) {
+    Rectangle tempRect1 = rect1;
+    Rectangle tempRect2 = rect2;
+
+    tempRect1.lowerLeftPoint = vec2I(rect1.lowerLeftPoint.x + rect1MovementDirection.x, rect1.lowerLeftPoint.y + rect1MovementDirection.y);
+    tempRect2.lowerLeftPoint = vec2I(rect2.lowerLeftPoint.x + rect2MovementDirection.x, rect1.lowerLeftPoint.y + rect2MovementDirection.y);
+
+    return checkCollision(&tempRect1, &tempRect2);

```

Comment:
Do you really need there to pass rectangles by reference ? I simply not sure is there would be any difference. Maybe reference passing wouldn't create a new object inside checkCollision scope. 

Anyway I think this solution is pretty good.

Reply:
Perhaps I don't fully understand how pointers work on C, but I don't want to change the state of the rectangles in this method. I mean, we need to get the collision that can happen if rect1 goes toward rect2